### PR TITLE
Implement useReactiveVar hook for consuming reactive variables in React components.

### PR DESCRIPTION
**Pull request by benjamn**
This PR implements a React hook called `useReactiveVar` (requested in #6818), which can be used to read from a [reactive variable](https://www.apollographql.com/blog/local-state-management-with-reactive-variables/) in a way that allows the React component to rerender if/when the variable is next updated.

Previously, the only way for a reactive variable to trigger a React component rerender was through the use of `useQuery`. As the tests included in this PR demonstrate, you don't have to be using `useQuery` (or GraphQL at all) to benefit from the reactivity that `ReactiveVar<T>` provides.

Implementing `useReactiveVar` required generalizing the `ReactiveVar<T>` API to allow arbitrary listeners (not just `InMemoryCache` objects), ~but fortunately the existing cache broadcast logic fits nicely within the new listener API~ (this unification was reverted after discussion with @glasser below).

### CODE REVIEW

**Comment by hwillson** [comment]
Just to add - we'll want to get some docs for this in place as well.

File: src/cache/inmemory/reactiveVars.ts [hunk 1]
```diff
@@ -12,34 +17,48 @@ const varDep = dep<ReactiveVar<any>>();
...
       if (value !== newValue) {
         value = newValue!;
         varDep.dirty(rv);
-        // Trigger broadcast for any caches that were previously involved
-        // in reading this variable.
-        caches.forEach(broadcast);
+        listeners.forEach(listener => {
+          // Listener functions listen only for the next update, not all
+          // future updates.
+          listeners.delete(listener);
```
Comment (glasser) [inline src/cache/inmemory/reactiveVars.ts:30]: This isn't a no-op change — I don't see old code that removed caches from `caches`. Was this change intentional? I don't see that you called it out in the description. I certainly believe it could be a bug fix if intentional!

Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts:30]: The difference here is that calling `broadcastWatches` on an `InMemoryCache` object is always safe/idempotent, whereas we can't make the same assumption about arbitrary listener functions.

Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts:30]: Also, adding the same `cache` object multiple times to the `caches` set would not lead to multiple broadcasts, but now that we're adding a new `() => broadcast(cache)` function each time, it's more important to remove functions as we call them.

Comment (glasser) [reply src/cache/inmemory/reactiveVars.ts:30]: Sure, but my concern is that it's not clear to me (a relatively uninformed person) just from reading the diff that nothing in the previous code *relied* on the fact that the cache would remain in the cache after the first broadcast call.  You're saying "it was OK to call it multiple times before"; my concern is "are we sure we will call it enough times if we're now removing it".

Again, I don't know the broader context so it's quite possible that my concern is baseless.

File: src/cache/inmemory/reactiveVars.ts [hunk 2]
```diff
@@ -12,34 +17,48 @@ const varDep = dep<ReactiveVar<any>>();
...
+        });
       }
     } else {
       // When reading from the variable, obtain the current cache from
       // context via cacheSlot. This isn't entirely foolproof, but it's
       // the same system that powers varDep.
       const cache = cacheSlot.getValue();
-      if (cache) caches.add(cache);
+      if (cache && (cache as any).broadcastWatches) {
+        listeners.add(() => broadcast(cache));
```
Comment (glasser) [inline src/cache/inmemory/reactiveVars.ts:40]: Do you still need the separate `broadcast` function, now that you're already checking that broadcastWatches exists? Seems like you could just do `listeners.add(() => cache.broadcastWatches())` here, right?

File: src/cache/inmemory/reactiveVars.ts (hunk 1 above)
Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts:30]: Thanks for pointing this out.

I think I could make an argument that calling `broadcastWatches` immediately causes rereading any affected queries, so "read, write, write" is actually "read, write, broadcast, read, write, broadcast, read," but I'm now leaning towards keeping the semantics the way they were, because I agree with your intuition that this change is somewhat risky—even if the risk is just that I might be missing something subtle that invalidates my optimistic argument.

I was also contemplating a change to avoid creating more than one `() => broadcast(cache)` callback per `cache` object, but that system would be more complicated than just maintaining a separate `Set<ApolloCache<any>>`, as we did before, so that's another reason to keep things as they were.

I am quite confident that `broadcastWatches` is idempotent, so calling it after any update to a reactive variable (possibly more often than necessary) is definitely safe.

Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts]: Done and tested: b874104cddf718547d9ff0b4fa51cda8267bbf2a

File: src/cache/inmemory/reactiveVars.ts (hunk 2 above)
Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts]: @glasser Since I ended up restoring the `Set<ApolloCache<any>>` system, I'm also going to keep the `broadcast` function, but you were probably right that it could have been removed!

File: docs/source/local-state/managing-state-with-field-policies.mdx [hunk 3]
```diff
@@ -147,15 +147,14 @@ This `read` function returns the value of our reactive variable whenever `cartIt
...
+```jsx{7}:title=AddToCartButton.js
 import { cartItemsVar } from './cache';
 // ... other imports
 
 export function AddToCartButton({ productId }) {
-  const cartItems = cartItemsVar();
   return (
     <div class="add-to-cart-button">
-      <Button onClick={() => cartItemsVar([...cartItems, productId])}>
+      <Button onClick={() => cartItemsVar([...cartItemsVar(), productId])}>
```
Comment (benjamn) [inline docs/source/local-state/managing-state-with-field-policies.mdx:157]: This was risky because the button could be clicked long after the component was first rendered.

File: docs/source/local-state/managing-state-with-field-policies.mdx [hunk 4]
```diff
@@ -199,7 +196,32 @@ export function Cart() {
 }
 ```
 
-</ExpansionPanel>
+Alternatively, you can read directly from a reactive variable using the `useReactiveVar` hook introduced in Apollo Client 3.2.0:
```
Comment (benjamn) [inline docs/source/local-state/managing-state-with-field-policies.mdx:199]: @StephenBarlow Is it okay to refer to an Apollo Client version number in the docs like this?

File: src/cache/inmemory/reactiveVars.ts [hunk 5]
```diff
@@ -3,25 +3,47 @@ import { dep } from "optimism";
...
 // Contextual Slot that acquires its value when custom read functions are
 // called in Policies#readField.
 export const cacheSlot = new Slot<ApolloCache<any>>();
 
+// A listener function could in theory cause another listener to be added
+// to the set while we're iterating over it, so it's important to commit
+// to the original elements of the set before we begin iterating. See
+// iterateObserversSafely for another example of this pattern.
+function consumeAndIterate<T>(set: Set<T>, callback: (item: T) => any) {
+  const items: T[] = [];
```
Comment (glasser) [inline src/cache/inmemory/reactiveVars.ts:24]: `const items = [...set]` ?

File: src/cache/inmemory/reactiveVars.ts [hunk 6]
```diff
@@ -34,12 +56,21 @@ export function makeVar<T>(value: T): ReactiveVar<T> {
 
     return value;
   };
+
+  rv.onNextChange = listener => {
+    listeners.add(listener);
+    return () => {
+      listeners.delete(listener);
```
Comment (glasser) [inline src/cache/inmemory/reactiveVars.ts:63]: Is there any concern that if somebody is passing a constant function to onNextChange (rather than one that has a new value every time onNextChange is called), that this might affect a different addition than the one done by this onNextChange call?

It's also a little odd in general that

```
rv.onNextChange(someFunc);
rv.onNextChange(someFunc);
```

is semantically different from

```
rv.onNextChange((v) => someFunc(v));
rv.onNextChange((v) => someFunc(v));
```

I think this is probably not a big deal and does not affect the useReactiveVar use case though.

Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts:63]: Since this is a new API, designed mostly for the sake of `useReactiveVar` (which passes fresh functions), I hope that people examine the code before using it, or at least don't blame us for their confusion otherwise.

File: src/cache/inmemory/reactiveVars.ts (hunk 5 above)
Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts:24]: You have every reason to expect that to work, but TypeScript complains:
```
Type 'Set<T>' is not an array type or a string type. Use compiler option '--downlevelIteration' to allow iterating of iterators. [2569]
```
The ugly truth is that `--downlevelIteration` adds a bunch of code to all `...spread` operations, so that's unfortunately not an option for us, until we're comfortable dropping support for Internet Explorer.
